Tests/SwiftProtobufTests/Test_SimpleExtensionMap.swift

```diff
@@ -0,0 +1,219 @@
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/master/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Test SimpleExtensionMap.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+import XCTest
+@testable import SwiftProtobuf
+
+extension AnyMessageExtension {
+  // Support equality to simplify testing of getting the correct errors.
+  func isEqual(_ other: AnyMessageExtension) -> Bool {
+    return (fieldNumber == other.fieldNumber &&
+      fieldName == other.fieldName &&
+      messageType == other.messageType)
+  }
+}
+
+// Define some extension to use for testing behaviors.
+
+let ext1 = MessageExtension<OptionalExtensionField<ProtobufInt32>, ProtobufUnittest_TestAllExtensions>(
+  _protobuf_fieldNumber: 1,
+  fieldName: "my_ext1"
+)
+
+let ext2 = MessageExtension<OptionalExtensionField<ProtobufInt64>, ProtobufUnittest_TestAllExtensions>(
+  _protobuf_fieldNumber: 2,
+  fieldName: "my_ext2"
+)
+
+// Same field number as ext1, but different class being extended.
+let ext3 = MessageExtension<OptionalExtensionField<ProtobufDouble>, ProtobufUnittest_TestPackedExtensions>(
+  _protobuf_fieldNumber: 1,
+  fieldName: "my_ext1b"
+)
+
+// Same field number and message type as ext2, so it will replace it in the mapping.
+let ext4 = MessageExtension<OptionalExtensionField<ProtobufBool>, ProtobufUnittest_TestAllExtensions>(
+  _protobuf_fieldNumber: 2,
+  fieldName: "my_ext4"
+)
+
+
+class Test_SimpleExtensionMap: XCTestCase {
+  func assert(map: SimpleExtensionMap, contains: [AnyMessageExtension], line: UInt = #line) {
+    // Extact what it constaings.
+    var includes = [AnyMessageExtension]()
+    for (_, l) in map.fields {
+      for e in l {
+        includes.append(e)
+      }
+    }
+
+    // Check that everything the lists match no matter the orders.
+    for c in contains {
+      var found = false
+      for i in includes {
+        if (c.isEqual(i)) {
+          found = true
+          break
+        }
+      }
+      XCTAssertTrue(found, "Map didn't include \(c)", line: line)
+    }
+    for i in includes {
+      var found = false
+      for c in contains {
+        if (i.isEqual(c)) {
+          found = true
+          break
+        }
+      }
+      XCTAssertTrue(found, "Map wasn't supposed to include \(i)", line: line)
+    }
+  }
+
+  func testInsert() {
+    var map = SimpleExtensionMap()
+    XCTAssertEqual(map.fields.count, 0)
+
+    map.insert(ext1)
+    assert(map: map, contains: [ext1])
+
+    map.insert(ext2)
+    assert(map: map, contains: [ext2, ext1])
+
+    map.insert(ext3)
+    assert(map: map, contains: [ext3, ext2, ext1])
+
+    map.insert(ext4)
```

Comment:
Added it to all the places so if someone has to look at any individual test case it will be clear.